Add `out` and `where` args for `ht.div`

neosunhan opened the pull request:
## Description
Implemention of `out` and `where` functionality for `ht.divide` is fairly straightforward. However, use of both arguments at the same time causes complications due to lack of support for in-place `where` in pytorch. 

Currently, I use a workaround by changing the underlying `torch.Tensor` of the `out` DNDarray. Possible alternatives include updating `indexing.where` to allow in-place modification or implementing a function that allows in-place modification using a mask (similar to [`numpy.copyto`](https://numpy.org/doc/stable/reference/generated/numpy.copyto.html)).
<!--- Include a summary of the change/s.
Please also include relevant motivation and context. List any dependencies that are required for this change.
--->

Issue/s resolved: #870

## Changes proposed:
- Use `out` functionality of `true_divide` in pytorch backend
- Use `indexing.where`


## Type of change
New feature (non-breaking change which adds functionality)
<!--
i.e.
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Documentation update
--->


## Due Diligence
- [x] All split configurations tested
- [ ] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

<!-- Remove this line for GPU Cluster tests. It will need an approval. --->

mtar commented:
GPU cluster tests are currently disabled on this Pull Request.

neosunhan commented:
Found a cleaner way to handle indexing of the `where` argument.

ClaudiaComito reviewed (changes requested):
@neosunhan thank you so much for taking this on!

I have a few comments, the main thing is that `where` should probably be addressed within `_operations.__binary_op()`. This way, it would be available to all binary operations. And it would be easier to satisfy the condition that `ht.divide(t1, t2)` returns `out`, not `t1`, where `where` is False (which also holds for all numpy binary operations).

Please let me know if you need help with `__binary_ops`!

neosunhan commented:
@ClaudiaComito Thanks for the feedback! I have taken a look at `_operations.__binary_op` and here is what I have come up with so far.
The main idea is to create a new DNDarray with uninitialized values (using `torch.empty`) in the event that `out=None`. We can then copy over the values from the result tensor (using `torch.where` to index if a condition is provided).

ClaudiaComito commented on heat/core/_operations.py:
```diff
@@ -163,23 +168,24 @@ def __get_out_params(target, other=None, map=None):
-        out.larray[:] = operation(
-            t1.larray.type(promoted_type), t2.larray.type(promoted_type), **fn_kwargs
+    else:
+        out_tensor = torch.empty(output_shape, dtype=promoted_type)
```
2 comments here:
- `output_shape` is the global (memory-distributed) shape of the output DNDarray, here you're initializing a potentially huge torch tensor. In this case you should call 
```
factories.empty(output_shape, dtype=..., split=..., device=...)
```
and that will take care of only initializing slices of the global array on each process

- if I understand the numpy docs correctly, this empty `out` DNDarray only needs to be initialized if `where` is not None.

ClaudiaComito reviewed (changes requested):
Great job @neosunhan ! :clap: I have left some comments in the code.

neosunhan commented:
@ClaudiaComito I have modified the code accordingly and added a few more tests.

ClaudiaComito commented:
run tests

ClaudiaComito reviewed (changes requested):
@neosunhan thanks a lot. This looks good, as far as I can tell. Can you update the CHANGELOG? When you're done, we can run the GPU tests.

neosunhan commented:
Updated CHANGELOG for the new `div` kwargs. I'm not sure which section is appropriate to document the new `where` kwarg for `__binary_op` (or if it should even be included in the changelog).

Also I believe there was a typo in the CHANGELOG where the "Feature Additions" heading was repeated 3 times, which I have fixed.

ClaudiaComito commented:
run tests

ClaudiaComito commented on heat/core/_operations.py:
```diff
@@ -43,6 +44,8 @@ def __binary_op(
     out: DNDarray, optional
         Output buffer in which the result is placed
+    where: DNDarray, optional
+        Condition of interest, where true yield the result of the operation else yield original value in out (uninitialized when out=None)
```
We can use numpy's docs for `where`, I think they are a bit clearer. But we must expand on them a bit, e.g. is `where` supposed/expected to be distributed, and how.

ClaudiaComito commented on heat/core/tests/test_arithmetics.py:
```diff
@@ -362,12 +362,42 @@ def test_div(self):
+        a = self.a_tensor.copy()
+        ht.div(a, self.a_scalar, out=a, where=ht.array([False, True]))
+        self.assertTrue(ht.equal(a, result_where_broadcasted))
+
```
Test case where `where` is distributed? Also the case in which `where_broadcasted` is distributed might get complicated.

ClaudiaComito reviewed (changes requested):
@neosunhan please bear with me. While we're trying to figure out why the GPU tests fail, it occurred to me that we are not checking for `where`'s distribution scheme and whether it fits the way `out` is distributed.

neosunhan commented:
@ClaudiaComito I added some tests where `out` and `where` are split differently and they all seem to pass. Not sure if I'm missing something here.

ClaudiaComito commented:
> 



> @ClaudiaComito I've added several test cases with different split configurations of `out` and `where`, but they all seem to pass. Not sure if I'm missing something here.

@neosunhan If you look at the list of checks below, you will see that the **continuous-integration/jenkins/pr-merge** has failed. Click on details, and you can see that the tests passed on 1 process, but failed on 2. Do you need help setting up your environment for multi-process? If you have installed `openmpi`, you should be able to run the tests on 2 processes with `mpirun -n 2 python -m unittest`.

neosunhan commented:
@ClaudiaComito Thanks for the help! I have added the check for `where`'s distribution scheme and the corresponding test cases.

ClaudiaComito commented:
run tests

ClaudiaComito commented on heat/core/_operations.py:
```diff
@@ -38,11 +39,17 @@ def __binary_op(
+        will be set to the result of the operation. Elsewhere, the `out` array will retain its original
+        value. If an uninitialized `out` array is created via the default `out=None`, locations within
+        it where the condition is False will remain uninitialized. If distributed, must be distributed
+        along the same dimension as the `out` array.
```
This is only correct if `where` and `out` have the same shape. 

For example if `out` is (100, 10000) and distributed along the columns (`out.split` is 1), `where` is (10000,), their shapes are broadcastable but `where` must be distributed along 0.

neosunhan replied:
>"If distributed, the split axis (after broadcasting if required) must match that of the `out` array."

Would this phrasing be accurate?

ClaudiaComito replied:
yes, sounds accurate

ClaudiaComito reviewed (changes requested):
I'm ready to approve this, great job @neosunhan . Just a small documentation update needed.

ClaudiaComito commented:
run tests

mtar reviewed (commented):
One check can be saved

ClaudiaComito commented:
run tests

ClaudiaComito reviewed (approved):
thanks a lot @neosunhan !
